### PR TITLE
fix: Catch silent IOU precision loss in vault/lending accounting

### PR DESCRIPTION
**Pull request by Tapanito**
Catch the bug class where IOU canonicalization (16-digit mantissa ceiling) on a vault or loan-broker accounting field rounds the recorded delta away from the requested amount. Pre-amendment this either fires `tecINVARIANT_FAILED` at finalize or, on the `accountSend` rails, slips past invariants entirely as silent value creation/destruction. Gated behind `fixCleanup3_2_0`.

## High-level overview

Three layers of defense (helper / preclaim / invariant) catch different shapes of the bug. One targeted invariant relaxation unsticks a vault when the only representable outcome is silent receiver-side absorption.

### 1. `accountSendExact` (TokenHelpers)

Drop-in replacement for `accountSend` that verifies the IOU transfer conserves value at the asset's precision. Three check shapes:

- **Two-sided** (the common case): both trust lines round independently; compare `senderDelta` to `receiverDelta` quantized at the coarser of the two pre-state scales. Sub-grid noise is admitted; ≥1 ULP discrepancy is rejected.
- **Mint** (sender is the asset issuer): only the receiver rounds; compare `receiverDelta` to the canonicalized requested amount.
- **Destroy** (receiver is the asset issuer): symmetric.

Skips entirely pre-amendment, for native and MPT (integer-exact), and for the degenerate issuer-to-self case. Returns `tecPRECISION_LOSS` on a real value-transfer mismatch.

### 2. Vault and LoanBroker helpers

Asset-side accounting extracted from each transactor:

- `depositToVault` / `withdrawFromVault` (VaultHelpers)
- `depositToBrokerCover` / `withdrawFromBrokerCover` (LendingHelpers)

Each helper snapshots pre-state, applies SLE updates + share/cover transfer + `associateAsset` + asset transfer (via `doWithdraw` for withdrawals), then cross-checks that the SLE accounting fields decremented in lockstep at the coarser of their pre-state scales — catches `associateAsset` asymmetric rounding on `sfAssetsTotal` vs `sfAssetsAvailable` (vault) or `sfCoverAvailable` vs broker pseudo trust line (broker). Issuer-as-counterparty carve-outs mirror the existing finalize-time invariants.

### 3. Preclaim sub-ULP guards

`VaultDeposit`, `VaultWithdraw`, `LoanBrokerCoverDeposit`, and `LoanBrokerCoverWithdraw` reject deposits/withdrawals whose amount is sub-ULP at the relevant accounting rail's pre-state scale — i.e. cases that the helper-level checks tolerate (both deltas round to zero) but which would otherwise leave SLE state inconsistent or trigger an invariant at finalize. Centralized in `rejectIfSubUlpAtCoarsestScale` (TokenHelpers).

### 4. VaultInvariant IOU absorption carve-out

The above checks reject super-ULP discrepancies, but a sole depositor whose own trust line sits past the 16-digit edge has no representable outcome other than canonicalization absorption: the vault loses the unit cleanly, the receiver's grid can't represent the inflow, the unit returns to the issuer's outstanding-obligation pool. Pre-amendment, `VaultInvariant`'s "withdrawal must increase destination balance" assertion fires at finalize, leaving the depositor structurally unable to exit the vault they fully owned.

The new IOU-only carve-out (`fixCleanup3_2_0` + `vaultAsset.holds<Issue>()`) tolerates a zero-rounded destination delta. The "by equal amount" check remains the safety net: vault loss > 1 ULP at the comparison scale paired with zero destination delta is still a genuine value-transfer mismatch and still fires. XRP and MPT remain strict — those types are integer-exact, no canonicalization, so a zero-rounded destination delta is unreachable for non-zero withdrawals.

Mirrors AMM withdraw semantics (silent absorption).

### Shared predicates

- `equalAtAssetScale(a, b, refA, refB, asset)` (STAmount) — quantize-then-equal at the coarser of the two reference scales.
- `firstNonzero(value, fallback)` (Number) — pick a representative magnitude from two observations of the same quantity when one may be true zero with a sentinel exponent.
- `rejectIfSubUlpAtCoarsestScale(asset, amount, references, context, j)` (TokenHelpers) — preclaim-style sub-ULP rejection used by all four transactors.

## Context of Change

Bug fix gated behind `fixCleanup3_2_0`. Pre-amendment behavior is preserved exactly: every check short-circuits or falls back, and the absorption carve-out is gated on both the amendment and the IOU asset type.

## API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change
- [x] \`libxrpl\` change — additive: \`accountSendExact\`, \`rejectIfSubUlpAtCoarsestScale\`, \`equalAtAssetScale\`, \`firstNonzero\`, vault/broker helpers
- [ ] Peer protocol change

## Test Plan

Bug-class regression anchors in `Vault_test.cpp`, `LoanBroker_test.cpp`, and `Loan_test.cpp` cover both amendment states (pre: bug reproduces; post: rejected with `tecPRECISION_LOSS`, except the absorption case which now succeeds):

- Two-sided edge crossings on vault deposit/withdraw (third-party and self-destination) and broker cover deposit/withdraw.
- Issuer-side mint and destroy at the edge (vault and broker cover).
- Sub-ULP requests against a high-magnitude pre-state (preclaim guard cases) for vault deposit/withdraw and broker cover deposit/withdraw.
- SLE-field delta-vs-request mismatch on `sfAssetsTotal` / `sfAssetsAvailable` / `sfLossUnrealized` (`testBugSleDeltaCheckTotalEdgeWithLoan`).
- Receiver-at-edge absorption (`testBugStuckVaultReceiverAtEdge`): pre-amendment fails with `tecINVARIANT_FAILED`; post-amendment succeeds with explicit post-state assertions — vault drained, alice's trust-line balance unchanged, shares burned.
- LoanSet origination-fee and disbursement edge cases.
- LoanManage impair edge case (parked, no fix applied — regression anchor for future work).

Unit tests for the new shared predicates: `testEqualAtAssetScale` (STAmount_test.cpp) and `testFirstNonzero` (Number_test.cpp).

### CODE REVIEW

File: src/libxrpl/ledger/helpers/VaultHelpers.cpp [hunk 1]
```diff
@@ -1,11 +1,26 @@
...
+#include <xrpl/ledger/helpers/TokenHelpers.h>
+#include <xrpl/protocol/AccountID.h>
+#include <xrpl/protocol/Asset.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
 #include <xrpl/protocol/SField.h>
 #include <xrpl/protocol/STAmount.h>
 #include <xrpl/protocol/STLedgerEntry.h>
 #include <xrpl/protocol/STNumber.h>  // IWYU pragma: keep
+#include <xrpl/protocol/STTakesAsset.h>
```
[the remark this answers is not in the document]
Comment (Tapanito) [reply src/libxrpl/ledger/helpers/VaultHelpers.cpp:18]: This brings in `associateAssets` call.

File: include/xrpl/protocol/STAmount.h [hunk 2]
```diff
@@ -749,6 +749,45 @@ scale(Number const& number, Asset const& asset)
...
+equalAtAssetScale(
+    Number const& a,
+    Number const& b,
+    Number const& referenceA,
+    Number const& referenceB,
+    Asset const& asset)
+{
+    int const sa = scale(referenceA, asset);
+    int const sb = scale(referenceB, asset);
+    int const minScale = sa > sb ? sa : sb;
```
[the remark this answers is not in the document]
Comment (Tapanito) [reply include/xrpl/protocol/STAmount.h]: renamed.

File: src/libxrpl/ledger/helpers/LendingHelpers.cpp [hunk 3]
```diff
@@ -1944,4 +1950,176 @@ loanMakePayment(
...
+            coverDelta,
+            pseudoDelta,
+            firstNonzero(coverBefore, coverAfter),
+            firstNonzero(pseudoBefore, pseudoAfter),
+            vaultAsset))
+    {
+        JLOG(j.error()) << "withdrawFromBrokerCover: sfCoverAvailable delta mismatch"
+                        << " coverDelta=" << to_string(coverDelta)
+                        << " pseudoDelta=" << to_string(pseudoDelta);
+        return tecPRECISION_LOSS;
```
Comment (pratikmankawde) [inline src/libxrpl/ledger/helpers/LendingHelpers.cpp]: I am seeing coverage warning in these three files. I think in this case we should address these warnings, since they are on the newly added paths.
- LendingHelpers.cpp,
- TokenHelpers.cpp,
- VaultHelpers.cpp

File: include/xrpl/basics/Number.h [hunk 4]
```diff
@@ -756,6 +756,29 @@ squelch(Number const& x, Number const& limit) noexcept
...
+ * temporaries for either argument without dangling-reference hazards.
+ *
+ * @param value     The primary observation.
+ * @param fallback  Returned when `value` is true zero.
+ * @return `value` if `value.signum() != 0`, else `fallback`.
+ */
+[[nodiscard]] constexpr Number
+firstNonzero(Number const& value, Number const& fallback) noexcept
+{
+    return value.signum() == 0 ? fallback : value;
```
Comment (pratikmankawde) [inline include/xrpl/basics/Number.h]: The signum function looks like this:
```cpp
int
signum() const noexcept
{
    if (negative_)
        return -1;
    return (mantissa_ != 0u) ? 1 : 0;
}
```
This returns `-1` for negative zero. So maybe the comment of `firstNonzero` should incorporate this?

File: include/xrpl/ledger/helpers/TokenHelpers.h [hunk 5]
```diff
@@ -273,6 +274,45 @@ accountSend(
...
+ *  "the requested amount is too small to be representable on this rail" as
+ *  a friendly tecPRECISION_LOSS instead of letting the operation reach
+ *  finalize and fire tecINVARIANT_FAILED.
+ *
+ *  The coarsest scale is `max(scale(ref, asset))` over `references`. If
+ *  `roundToAsset(asset, amount, coarsestScale).signum() == 0`, returns
+ *  tecPRECISION_LOSS with `context` in the warn log; otherwise tesSUCCESS.
+ */
+[[nodiscard]] TER
+rejectIfSubUlpAtCoarsestScale(
```
Comment (pratikmankawde) [inline include/xrpl/ledger/helpers/TokenHelpers.h]: This will imply a new rule. Because earlier sub-ULP requests were allowed, (although resulting in precision losses). But is there a chance someone was still using it, since it still did something to their overall portfolio(not sure if this is a right word)? And now they can't.

File: include/xrpl/basics/Number.h [hunk 6]
```diff
@@ -756,6 +756,29 @@ squelch(Number const& x, Number const& limit) noexcept
...
+ *
+ * Returns by value (not by const reference) so callers may pass
+ * temporaries for either argument without dangling-reference hazards.
+ *
+ * @param value     The primary observation.
+ * @param fallback  Returned when `value` is true zero.
+ * @return `value` if `value.signum() != 0`, else `fallback`.
+ */
+[[nodiscard]] constexpr Number
+firstNonzero(Number const& value, Number const& fallback) noexcept
```
Comment (kuznetsss) [inline include/xrpl/basics/Number.h]: Shouldn't it be a method of `Number`?
```cpp
constexpr Number
firstNonzeroOr(Number const& fallback) noexcept;
```

File: src/libxrpl/ledger/helpers/LendingHelpers.cpp [hunk 7]
```diff
@@ -1944,4 +1950,176 @@ loanMakePayment(
...
+depositToBrokerCover(
+    ApplyView& view,
+    std::shared_ptr<SLE> const& broker,
+    AccountID const& depositor,
+    STAmount const& amount,
+    beast::Journal j)
+{
+    auto const vault = view.read(keylet::vault(broker->at(sfVaultID)));
+    if (!vault)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
```
Comment (kuznetsss) [inline src/libxrpl/ledger/helpers/LendingHelpers.cpp:2098]: Is it not possible to pass here a `broker` without `sfVaultID` field?
Should we add an assert before this return?

File: include/xrpl/protocol/STAmount.h [hunk 8]
```diff
@@ -749,6 +749,39 @@ scale(Number const& number, Asset const& asset)
...
+ *   - Zero IOU canonicalizes with a sentinel exponent (-100) and yields
+ *     no useful precision; pass the first non-zero observation of that
+ *     endpoint instead.
+ *
+ * For native (XRP) and MPT, scale is always 0 → comparison is at integer
+ * precision → predicate degenerates to strict equality (which is exact
+ * for those asset types).
+ */
+[[nodiscard]] bool
+equalAtAssetScale(
```
Comment (pratikmankawde) [inline include/xrpl/protocol/STAmount.h:802]: Any plans of consolidating `equalAtAssetScale` with  `VaultInvariant::computeCoarsestScale + roundToAsset`?

File: include/xrpl/ledger/helpers/TokenHelpers.h (hunk 5 above)
Comment (pratikmankawde) [reply include/xrpl/ledger/helpers/TokenHelpers.h]: If a vault contain sub-ULP amounts, it can never be emptied and can never be deleted. What's the solution in this case?

Comment (Tapanito) [reply include/xrpl/ledger/helpers/TokenHelpers.h]: I did not consider this edge case @pratikmankawde good call. Let me think it through if such a case is possible.

File: src/libxrpl/ledger/helpers/LendingHelpers.cpp (hunk 7 above)
Comment (Tapanito) [reply src/libxrpl/ledger/helpers/LendingHelpers.cpp:2098]: That's not possible, but it is possible to pass any `sle` which might be not `ltLOAN_BROKER`! I added an assertion higher up, to ensure that correct SLE is passed.

File: src/libxrpl/tx/transactors/lending/LoanBrokerCoverDeposit.cpp [hunk 9]
```diff
@@ -4,11 +4,11 @@
 #include <xrpl/beast/utility/Zero.h>
 #include <xrpl/ledger/helpers/LendingHelpers.h>
```
Comment (pratikmankawde) [inline src/libxrpl/tx/transactors/lending/LoanBrokerCoverDeposit.cpp:5]: Do we not need to cover the **LoanBrokerCoverClawback** case?

File: src/test/protocol/STAmount_test.cpp [hunk 10]
```diff
@@ -1209,6 +1209,80 @@ class STAmount_test : public beast::unit_test::Suite
 
     //--------------------------------------------------------------------------
 
+    void
+    testEqualAtAssetScale()
+    {
+        testcase("equalAtAssetScale");
+
+        Asset const xrp{xrpIssue()};
+        Asset const usd{Issue{Currency{1}, AccountID(0x4985601)}};
```
Comment (pratikmankawde) [inline src/test/protocol/STAmount_test.cpp:1218]: Will it make sense to `constexpr` the account id at a common place?

File: src/test/protocol/STAmount_test.cpp [hunk 11]
```diff
@@ -1209,6 +1209,80 @@ class STAmount_test : public beast::unit_test::Suite
...
+            Number const ref{1'000'000};
+            BEAST_EXPECT((equalAtAssetScale(Number{5}, Number{5}, ref, ref, xrp)));
+            BEAST_EXPECT((!equalAtAssetScale(Number{5}, Number{4}, ref, ref, xrp)));
+        }
+
+        // MPT: scale = 0 always, integer equality.
+        {
+            Number const ref{1'000'000};
+            BEAST_EXPECT((equalAtAssetScale(Number{5}, Number{5}, ref, ref, mpt)));
+            BEAST_EXPECT((!equalAtAssetScale(Number{5}, Number{4}, ref, ref, mpt)));
```
Comment (pratikmankawde) [inline src/test/protocol/STAmount_test.cpp:1233]: Do we need tests for:
- Zero ref. `equalAtAssetScale(a, b, Number{0}, <someref/Number{0}>, asset);`?
- Zero deltas: `equalAtAssetScale(Number{0}, Number{0}, ref, ref, asset)`, `equalAtAssetScale(Number{0}, Number{1}, ref, ref, asset)`
- Negative numbers or mixed signs: `equalAtAssetScale(Number{-2}, Number{-1}, refBob, refVault, usd);`
- Exact ULP: Vault at 10^16 (ULP=10), delta of exactly 10: `equalAtAssetScale(Number{10}, Number{10}, refSmall, refVault, usd)`
- Diff. rounding modes.

File: src/test/protocol/STAmount_test.cpp [hunk 12]
```diff
@@ -1209,6 +1209,80 @@ class STAmount_test : public beast::unit_test::Suite
...
+        }
+
+        // IOU same-magnitude with sub-grid noise (must accept). This is
+        // the false-positive case from the precision-loss fix: sender at
+        // ~10^6 (scale -9), receiver at ~10^3 (scale -12). minScale = -9.
+        // Both deltas round to 4999.50990099 at scale -9.
+        {
+            Number const refSender{1'009'901};
```
Comment (pratikmankawde) [inline src/test/protocol/STAmount_test.cpp:1241]: Maybe add tests for larger magnitude differences?
Sender ref: Number{1, -12} (0.000000000001 — a dust holder)
Receiver ref: Number{1, 15} (10^15 — at the IOU edge)
Scale gap: ~27
Edge case:
```cpp
equalAtAssetScale(Number{1, -85},          // near Number's minimum
                  Number{2, -85},          // near Number's minimum
                  Number{1, 15},           // near IOU maximum
                  Number{1, 15},           // near IOU maximum
                  usd);
```

File: src/test/protocol/STAmount_test.cpp [hunk 13]
```diff
@@ -1209,6 +1209,80 @@ class STAmount_test : public beast::unit_test::Suite
...
+        }
+
+        // IOU boundary-crossing 1-ULP (must reject). The bug class:
+        // depositor at low magnitude (Bob, scale -12) loses 2; vault at
+        // 16-digit edge (scale 0) gains 1. minScale = 0, so deltas round
+        // to 2 vs 1.
+        {
+            Number const refBob{1'000};
+            Number const refVault{9'999'999'999'999'999LL};
+            BEAST_EXPECT((!equalAtAssetScale(Number{2}, Number{1}, refBob, refVault, usd)));
```
Comment (pratikmankawde) [inline src/test/protocol/STAmount_test.cpp:1256]: Another potential test: `equalAtAssetScale(Number{1}, Number{6, -1}, Number{6, -1}, Number{6, -1}, usd)`

**Review by pratikmankawde — COMMENTED** [review]
Added few comments

File: include/xrpl/protocol/STAmount.h (hunk 8 above)
Comment (pratikmankawde) [inline include/xrpl/protocol/STAmount.h:802]: Can `equalAtAssetScale` be `noexcept`? Or catch the `overflow_error`?

Comment (Tapanito) [reply include/xrpl/protocol/STAmount.h:802]: I did think of that, but I think at this point it'd be out of scope for this PR. It was way too big as it is.

File: src/libxrpl/tx/transactors/lending/LoanBrokerCoverDeposit.cpp (hunk 9 above)
Comment (Tapanito) [reply src/libxrpl/tx/transactors/lending/LoanBrokerCoverDeposit.cpp:5]: Good call, added.

File: src/libxrpl/tx/transactors/lending/LoanSet.cpp [hunk 14]
```diff
@@ -559,14 +558,28 @@ LoanSet::doApply()
...
+    {
+        if (auto const ter = accountSendExact(
+                view,
+                vaultPseudo,
+                brokerOwner,
+                STAmount{vaultAsset, originationFee},
+                j_,
+                WaiveTransferFee::Yes))
+            return ter;
+    }
```
Comment (Tapanito) [inline src/libxrpl/tx/transactors/lending/LoanSet.cpp]: @ximinez, what do you make of this change, instead of the multi-send to use two exact sends?

Comment (Tapanito) [reply src/libxrpl/tx/transactors/lending/LoanSet.cpp]: Actually, I answered my own question: it's out of scope for this PR.

File: include/xrpl/basics/Number.h (hunk 4 above)
Comment (Tapanito) [reply include/xrpl/basics/Number.h]: Done.

File: include/xrpl/ledger/helpers/TokenHelpers.h (hunk 5 above)
Comment (Tapanito) [reply include/xrpl/ledger/helpers/TokenHelpers.h]: Done. This is addressed for withdrawals. When withdrawing the funds are clamped to the larges representable amount that is less that asked amount. If such amount does not exist, funds cannot be withdrawn, user will have to wait until vault scale is such that their withdrawal can be represented.

File: include/xrpl/basics/Number.h (hunk 6 above)
Comment (Tapanito) [reply include/xrpl/basics/Number.h]: Made it a method of Number.

File: src/libxrpl/basics/Number.cpp [hunk 15]
```diff
@@ -835,6 +835,12 @@ Number::truncate() const noexcept
     return ret;
 }
 
+[[nodiscard]] Number
```
Comment (a1q123456) [inline src/libxrpl/basics/Number.cpp:838]: Maybe you can return a Number const& to avoid copying

**Comment by Tapanito** [comment]
Closing in favor of smaller PRs:

* https://github.com/XRPLF/rippled/pull/7274 
* https://github.com/XRPLF/rippled/pull/7272
